tensorflow/core/kernels/mkl_avgpooling_op.cc

```diff
@@ -467,65 +466,62 @@ class MklAvgPoolingOp : public MklPoolingForwardOpBase<T> {
-        output_mkl_shape.SetMklTensor(false);
-        TensorShape output_tf_shape;
-        if (pool_params.data_format == TensorFormat::FORMAT_NCHW) {
-          output_tf_shape = MklDnnDimsToTFShape(output_dims_mkl_order);
-        } else {
-          memory::dims output_dims_NHWC_order;
-          output_dims_NHWC_order = {pool_params.tensor_in_batch,
-                                    static_cast<int>(pool_params.out_height),
-                                    static_cast<int>(pool_params.out_width),
-                                    pool_params.out_depth};
-          output_tf_shape = MklDnnDimsToTFShape(output_dims_NHWC_order);
-        }
         const int kOutputIndex = 0;
-        AllocateOutputSetMklShape(context, kOutputIndex, &output_tensor,
-                                  output_tf_shape, output_mkl_shape);
-        CHECK_NOTNULL(output_tensor);
+        this->AllocateEmptyOutputTensor(context, kOutputIndex, &pool_params,
+                                        output_dims_mkl_order, &output_tensor);
         return;
       }
 
-      // If input is in Mkl layout, then just get the memory format from it
-      // directly, instead of using input data_format to AvgPool.
-      if (dnn_shape_input.IsMklTensor()) {
-        dnn_data_output.SetUsrMem(
-            output_dims_mkl_order,
-            static_cast<memory::format>(
-                dnn_data_input.GetUsrMemDesc().data.format));
+      memory::dims filter_dims, strides, padding_left, padding_right;
+      this->PoolParamsToDims(&pool_params, &filter_dims, &strides,
+          &padding_left, &padding_right);
+
+      // Get the input memory descriptor
+      memory::desc input_md = dnn_shape_input.IsMklTensor()
+          ? dnn_shape_input.GetMklLayout()
+          : memory::desc(TFShapeToMklDnnDimsInNCHW(input_tensor_shape,
+                this->data_format_tf_),
+                MklDnnType<T>(), this->data_format_mkldnn_);
+
+      // Get src/filter/stride/padding information
+      memory::dims src_dims = dnn_shape_input.IsMklTensor()
+                            ? dnn_shape_input.GetSizesAsMklDnnDims()
+                            : TFShapeToMklDnnDimsInNCHW(input_tensor.shape(),
+                                   this->data_format_tf_);
+
+      // Get an average pooling primitive from the op pool
+      MklPoolingFwdPrimitive<T> *pooling_fwd = nullptr;
+      MklPoolingParams fwdParams(src_dims, output_dims_mkl_order, filter_dims,
+          strides, padding_left, padding_right,
+          algorithm::pooling_avg_exclude_padding);
+      pooling_fwd = MklPoolingFwdPrimitiveFactory<T>::Get(fwdParams);
+
+      // allocate output tensor
+      this->AllocateOutputTensor(context, *(pooling_fwd->GetPoolingFwdPd()),
+          output_dims_mkl_order, this->data_format_mkldnn_, &output_tensor);
+      CHECK_NOTNULL(output_tensor);
+
+      OP_REQUIRES_OK(context, context->status());
 
+      // check whether we need to reorder src
+      T* src_data = nullptr;
```

Comment:
Make this be "const T*". Perhaps simplify code a bit by just doing something

const T* src_data = input_tensor.flat<T>().data();
if (input_md.data.format != pooling_fwd->GetSrcMemoryFormat()) {
   ...
   // overwrite src_data
}

Reply:
Done with code change